control-plane/api-gateway/gatekeeper/rolebinding.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package gatekeeper
+
+import (
+	"context"
+	"errors"
+
+	"k8s.io/apimachinery/pkg/types"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/hashicorp/consul-k8s/control-plane/api-gateway/common"
+	"github.com/hashicorp/consul-k8s/control-plane/api/v1alpha1"
+	rbac "k8s.io/api/rbac/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+func (g *Gatekeeper) upsertRoleBinding(ctx context.Context, gateway gwv1beta1.Gateway, gcc v1alpha1.GatewayClassConfig, config common.HelmConfig) error {
+	if config.AuthMethod == "" {
+		return g.deleteRole(ctx, types.NamespacedName{Namespace: gateway.Namespace, Name: gateway.Name})
+	}
+
+	roleBinding := &rbac.RoleBinding{}
+
+	// If the RoleBinding already exists, ensure that we own the RoleBinding
+	err := g.Client.Get(ctx, g.namespacedName(gateway), roleBinding)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	} else if !k8serrors.IsNotFound(err) {
+		// Ensure we own the Role.
+		for _, ref := range roleBinding.GetOwnerReferences() {
+			if ref.UID == gateway.GetUID() && ref.Name == gateway.GetName() {
+				// We found ourselves!
+				return nil
+			}
+		}
+		return errors.New("role not owned by controller")
+	}
+
+	// Create or update the RoleBinding
+	roleBinding = g.roleBinding(gateway, gcc, config)
+	if err := ctrl.SetControllerReference(&gateway, roleBinding, g.Client.Scheme()); err != nil {
+		return err
+	}
+	if err := g.Client.Create(ctx, roleBinding); err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
Not blocking, but this looks almost identical to the role upsert logic, would it be worth considering refactoring these into a generic upsert implementation?

Reply:
It's almost identical for every resource we're creating. I'd love to refactor but don't wanna blow up the scope of this PR. We can file some work in the current epic to look into this if you're okay with that.